lib/geoip.rb

```diff
@@ -914,22 +914,21 @@ def le_to_ui(s) #:nodoc:
   # access (only providing protection against multiple threads, but not
   # file descriptors shared across multiple processes).
   # If the contents of the database have been preloaded it'll work with
-  # the StringIO object directly.
+  # the frozen string object directly.
   def atomic_read(length, pos) #:nodoc:
-    if @mutex
+    if @contents
+      @contents.byteslice(pos, length) || ''
+    elsif @use_pread
+      IO.pread(@file.fileno, length, pos)
+    elsif @mutex
       @mutex.synchronize { atomic_read_unguarded(length, pos) }
     else
       atomic_read_unguarded(length, pos)
     end
   end
 
   def atomic_read_unguarded(length, pos)
```

Comment:
This routine is no longer atomic, so that should be removed from its name.

Reply:
Done.